### PR TITLE
Phase A: Shuttle Integration - Project ID and Shuttle Assignment

### PR DESCRIPTION
**Pull request by mithro**
## Summary

Implements **Phase A** of the shuttle integration feature, adding shuttle assignment and project ID functionality to the wafer.space platform.

### Features Added

**1. Shuttle Assignment**
- Projects can now be assigned to specific shuttle runs (e.g., G801, G802)
- Shuttle selection dropdown in project form
- Shuttle information displayed on project detail page

**2. Project IDs**
- 4-character alphanumeric project IDs (A-Z, 0-9, uppercase)
- Unique per shuttle (different shuttles can reuse same ID)
- Real-time validation with live feedback

**3. Manufacturing IDs**
- 8-character full IDs combining shuttle + project ID (e.g., "G801ABCD")
- Displayed prominently on project detail page

**4. Shuttle ID Format Validation**
- Enforces strict format: G8XX (where XX are digits 00-99)
- Valid examples: G800, G801, ..., G899

**5. Enhanced UX**
- Bootstrap popovers with help text
- Live character counting (X/4 characters)
- AJAX-based availability checking
- Automatic uppercase conversion and alphanumeric filtering
- Dynamic slot size filtering based on selected shuttle

### Technical Implementation

**New Infrastructure**
- Created `wafer_space/core` module with shared `SlotSize` enum
- Added shuttle FK and project_id fields to Project model
- Comprehensive validation at form, model, and database levels

**Database Changes**
- 3 migrations created (projects.0039, 0040; shuttles.0002)
- Unique constraint on (shuttle, project_id) combination
- Database indexes for query performance

**Backend**
- AJAX endpoints for real-time validation
  - `/projects/check-project-id/` - validates availability
  - `/projects/shuttle-available-sizes/` - returns slot sizes
- Validator functions with constants (no magic numbers)

**Frontend**
- JavaScript for real-time validation with debouncing
- Character count display and input filtering
- Enhanced project detail and form templates

**Admin Interface**
- Updated ProjectAdmin with shuttle, project_id, and full_id columns
- Search and filter capabilities

### Files Changed

- **Created:** `wafer_space/core/` module
- **Modified:** 15 files across models, views, forms, templates, tests
- **Migrations:** 3 new migrations
- **Tests:** Updated and all passing ✅

### Code Quality

- ✅ All tests passing (559 tests)
- ✅ Linting (ruff) passing
- ✅ Type checking (mypy) passing
- ✅ 16 commits with detailed messages
- ✅ No magic numbers (constants used throughout)

### Test Coverage

- Updated existing ProjectForm tests
- All 559 tests in projects app passing
- Validation tested with edge cases

### Next Steps: Phase B

Phase B will add:
- Grid coordinate fields to ShuttleSlot (row, column)
- Shuttle configuration with row_heights/column_widths
- Staff-only shuttle admin dashboard
- Slot assignment interface with size mismatch warnings

---

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

## Summary by CodeRabbit

* **New Features**
  * Projects can be linked to shuttles, receive 4‑char project IDs, and use shared slot-size choices; live availability checks and shuttle-specific size options in the form.

* **Admin**
  * Project admin list, filters, and search expanded to show shuttle and project ID.

* **UI**
  * Project detail and form UI: conditional shuttle/project ID display, updated date formats, popover help, uppercase enforcement, and real‑time validation.

* **Tests**
  * Extensive tests for validators, shuttle/project integration, forms, AJAX endpoints, tasks, and admin flows.

* **Chores**
  * Data migrations add and populate shuttle and project ID fields; shuttle ID validation and description added.

* **Documentation**
  * Module docstrings and a shuttle grid layout design document added.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: wafer_space/projects/migrations/0041_populate_shuttle_and_project_ids.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    """Reverse the data migration by clearing shuttle and project_id."""
+    Project = apps.get_model("projects", "Project")
+
+    # Clear shuttle and project_id for all projects
+    projects = Project.objects.filter(shuttle__isnull=False)
+    count = projects.count()
+
+    projects.update(shuttle=None, project_id="")
+
+    print(f"Cleared shuttle and project_id for {count} projects")
```
[the remark this answers is not in the document]
Comment (mithro) [reply wafer_space/projects/migrations/0041_populate_shuttle_and_project_ids.py]: Fixed in commit 9352c62 - replaced all `print()` statements with `logger.info()` using Python logging module.

File: wafer_space/projects/tasks.py [hunk 2]
```diff
@@ -561,12 +561,14 @@ def _start_container(context: _CheckContext) -> tuple[str, str]:
         top_cell,
         "--slot",
         slot_size,
+        "--id",
+        context.project.full_id,
```
[the remark this answers is not in the document]
Comment (mithro) [reply wafer_space/projects/tasks.py]: Fixed in commit fe2d127 - added validation to ensure `full_id` is not empty before building the docker command.
